File: src/lib/stats.ts
export type StatEvent = {
  type: string;
  startedAt: Date;
  endedAt: Date | null;
  details: Record<string, unknown>;
};

export type DayStats = {
  date: string;
  sleepMinutes: number;
  breastMinutes: number;
  feedCount: number;
  bottleMl: number;
  diaperWet: number;
  diaperDirty: number;
  diaperBoth: number;
  pumpMl: number;
  medicineCount: number;
};

const DAY_MS = 86_400_000;

function dayKey(d: Date, tzOffsetMinutes: number): string {
  return new Date(d.getTime() - tzOffsetMinutes * 60_000).toISOString().slice(0, 10);
}

function emptyDay(date: string): DayStats {
  return {
    date, sleepMinutes: 0, breastMinutes: 0, feedCount: 0, bottleMl: 0,
    diaperWet: 0, diaperDirty: 0, diaperBoth: 0, pumpMl: 0, medicineCount: 0,
  };
}

/** End of the local day containing `d`, as a UTC instant. */
function nextLocalMidnight(d: Date, tzOffsetMinutes: number): Date {
  const local = new Date(d.getTime() - tzOffsetMinutes * 60_000);
  const midnightUtcOfNextLocalDay = Date.UTC(
    local.getUTCFullYear(), local.getUTCMonth(), local.getUTCDate() + 1,
  );
  return new Date(midnightUtcOfNextLocalDay + tzOffsetMinutes * 60_000);
}

export function aggregateDaily(
  eventsIn: StatEvent[], days: number, tzOffsetMinutes: number, now: Date,
): DayStats[] {
  const byDay = new Map<string, DayStats>();
  for (let i = days - 1; i >= 0; i--) {
    const key = dayKey(new Date(now.getTime() - i * DAY_MS), tzOffsetMinutes);
    byDay.set(key, emptyDay(key));
  }

  for (const e of eventsIn) {
    // Durations (sleep + breast feeds) are spread across local days.
    const isBreast = e.type === "feed" && e.details?.method === "breast";
    if (e.type === "sleep" || isBreast) {
      const end = e.endedAt ?? now;
      let cursor = e.startedAt;
      while (cursor < end) {
        const boundary = nextLocalMidnight(cursor, tzOffsetMinutes);
        const sliceEnd = end < boundary ? end : boundary;
        const minutes = (sliceEnd.getTime() - cursor.getTime()) / 60_000;
        const day = byDay.get(dayKey(cursor, tzOffsetMinutes));
        if (day) {
          if (e.type === "sleep") day.sleepMinutes += minutes;
          else day.breastMinutes += minutes;
        }
        cursor = sliceEnd;
      }
    }

    // Counts/volumes attributed to the start day.
    const day = byDay.get(dayKey(e.startedAt, tzOffsetMinutes));
    if (!day) continue;
    const det = e.details ?? {};
    switch (e.type) {
      case "feed": {
        day.feedCount += 1;
        if (det.method === "bottle" && typeof det.amountMl === "number") day.bottleMl += det.amountMl;
        break;
      }
      case "diaper": {
        if (det.kind === "wet") day.diaperWet += 1;
        else if (det.kind === "dirty") day.diaperDirty += 1;
        else if (det.kind === "both") day.diaperBoth += 1;
        break;
      }
      case "pump": {
        const l = typeof det.leftMl === "number" ? det.leftMl : 0;
        const r = typeof det.rightMl === "number" ? det.rightMl : 0;
        day.pumpMl += l + r;
        break;
      }
      case "medicine": day.medicineCount += 1; break;
    }
  }

  const out = [...byDay.values()];
  for (const d of out) {
    d.sleepMinutes = Math.round(d.sleepMinutes);
    d.breastMinutes = Math.round(d.breastMinutes);
  }
  return out;
}

File: backend/app/stats.py
"""Daily aggregation of events, a direct port of the former TypeScript lib/stats.ts."""

import math
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any

from app.mcp.summaries import merge_intervals

DAY = timedelta(days=1)


def _js_round(x: float) -> int:
    """JS Math.round semantics (halves round up), unlike Python's banker's rounding."""
    return math.floor(x + 0.5)


@dataclass
class StatEvent:
    type: str
    started_at: datetime
    ended_at: datetime | None
    details: dict[str, Any] = field(default_factory=dict)


def _day_key(d: datetime, tz_offset_minutes: int) -> str:
    return (d - timedelta(minutes=tz_offset_minutes)).astimezone(timezone.utc).date().isoformat()


def _empty_day(date_str: str) -> dict[str, Any]:
    return {
        "date": date_str,
        "sleepMinutes": 0.0,
        "breastMinutes": 0.0,
        "feedCount": 0,
        "bottleMl": 0,
        "diaperWet": 0,
        "diaperDirty": 0,
        "diaperBoth": 0,
        "pumpMl": 0,
        "medicineCount": 0,
    }


def _next_local_midnight(d: datetime, tz_offset_minutes: int) -> datetime:
    """End of the local day containing `d`, as a UTC instant."""
    local = (d - timedelta(minutes=tz_offset_minutes)).astimezone(timezone.utc)
    next_midnight = datetime(
        local.year, local.month, local.day, tzinfo=timezone.utc
    ) + DAY
    return next_midnight + timedelta(minutes=tz_offset_minutes)


def aggregate_daily(
    events: list[StatEvent], days: int, tz_offset_minutes: int, now: datetime
) -> list[dict[str, Any]]:
    by_day: dict[str, dict[str, Any]] = {}
    for i in range(days - 1, -1, -1):
        key = _day_key(now - i * DAY, tz_offset_minutes)
        by_day[key] = _empty_day(key)

    # Sleep intervals are collected per local day (clipped to day boundaries)
    # and merged (union) below, rather than summed directly, so overlapping
    # sleep events don't double-count a day's total.
    sleep_intervals_by_day: dict[str, list[tuple[datetime, datetime]]] = defaultdict(list)

    for e in events:
        details = e.details or {}
        is_breast = e.type == "feed" and details.get("method") == "breast"

        # Durations (sleep + breast feeds) are spread across local days.
        if e.type == "sleep" or is_breast:
            end = e.ended_at or now
            cursor = e.started_at
            while cursor < end:
                boundary = _next_local_midnight(cursor, tz_offset_minutes)
                slice_end = min(end, boundary)
                day_key = _day_key(cursor, tz_offset_minutes)
                if e.type == "sleep":
                    if day_key in by_day:
                        sleep_intervals_by_day[day_key].append((cursor, slice_end))
                else:
                    minutes = (slice_end - cursor).total_seconds() / 60
                    bucket = by_day.get(day_key)
                    if bucket is not None:
                        bucket["breastMinutes"] += minutes
                cursor = slice_end

        # Counts/volumes attributed to the start day.
        bucket = by_day.get(_day_key(e.started_at, tz_offset_minutes))
        if bucket is None:
            continue
        if e.type == "feed":
            bucket["feedCount"] += 1
            amount = details.get("amountMl")
            if details.get("method") == "bottle" and isinstance(amount, (int, float)):
                bucket["bottleMl"] += amount
        elif e.type == "diaper":
            kind = details.get("kind")
            if kind == "wet":
                bucket["diaperWet"] += 1
            elif kind == "dirty":
                bucket["diaperDirty"] += 1
            elif kind == "both":
                bucket["diaperBoth"] += 1
        elif e.type == "pump":
            left = details.get("leftMl")
            right = details.get("rightMl")
            bucket["pumpMl"] += (left if isinstance(left, (int, float)) else 0) + (
                right if isinstance(right, (int, float)) else 0
            )
        elif e.type == "medicine":
            bucket["medicineCount"] += 1

    for day_key, intervals in sleep_intervals_by_day.items():
        bucket = by_day[day_key]
        merged = merge_intervals(intervals)
        bucket["sleepMinutes"] = sum((e - s).total_seconds() / 60 for s, e in merged)

    out = list(by_day.values())
    for d in out:
        d["sleepMinutes"] = _js_round(d["sleepMinutes"])
        d["breastMinutes"] = _js_round(d["breastMinutes"])
    return out
